src/sync/mutex.rs

```diff
@@ -0,0 +1,475 @@
+/// # */
+/// #    fn new() -> Self {
+/// #        Self {
+/// #            map1: HashMap::new(),
+/// #            map2: HashMap::new(),
+/// #        }
+/// #    }
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mutex = CMutex::new(MyStruct::new());
+///
+///     let mut permit = mutex.lock().await.unwrap();  // note unwrap() here
+///     permit.perform(|data| {
+///         data.map1.insert("hello".to_owned(), "world".to_owned());  // (1)
+///         // ... some code in between
+///         data.map2.insert("hello".to_owned(), 42);  // (2)
+///     });
+/// }
+/// ```
+///
+/// # Features
+///
+/// Basic mutex operations are supported. In the future, this will support:
+///
+/// - An `OwnedActionPermit`, similar to [`tokio::sync::OwnedMutexGuard`].
+/// - A `MappedActionPermit`, similar to [`tokio::sync::MappedMutexGuard`].
+pub struct CMutex<T: ?Sized> {
+    poison: poison::Flag,
+    inner: tokio::sync::Mutex<T>,
+}
+
+impl<T: ?Sized> CMutex<T> {
+    /// Creates a new lock in an unlocked state ready for use.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use cancel_safe_futures::sync::CMutex;
+    ///
+    /// let lock = CMutex::new(5);
+    /// ```
+    #[track_caller]
+    pub fn new(value: T) -> Self
+    where
+        T: Sized,
+    {
+        Self {
+            inner: tokio::sync::Mutex::new(value),
+            poison: poison::Flag::new(),
+        }
+    }
+
+    /// Creates a new lock in an unlocked state ready for use.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use cancel_safe_futures::sync::CMutex;
+    ///
+    /// static LOCK: CMutex<i32> = CMutex::const_new(5);
+    /// ```
+    #[cfg(all(feature = "parking_lot", not(all(loom, test)),))]
+    #[cfg_attr(doc_cfg, doc(cfg(feature = "parking_lot")))]
+    pub const fn const_new(value: T) -> Self
+    where
+        T: Sized,
+    {
+        Self {
+            inner: tokio::sync::Mutex::const_new(value),
+            poison: poison::Flag::new(),
+        }
+    }
+
+    /// Locks this mutex, causing the current task to yield until the lock has been acquired.  When
+    /// the lock has been acquired, function returns a [`ActionPermit`].
+    ///
+    /// # Errors
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then this call will return
+    /// an error once the mutex is acquired.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method uses a queue to fairly distribute locks in the order they were requested.
+    /// Cancelling a call to `lock` makes you lose your place in the queue.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use cancel_safe_futures::sync::CMutex;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mutex = CMutex::new(1);
+    ///
+    ///     let mut permit = mutex.lock().await.unwrap();
+    ///     permit.perform(|n| *n = 2);
+    /// }
+    /// ```
+    pub async fn lock(&self) -> LockResult<ActionPermit<'_, T>> {
```

Comment:
is there a reason not to combine the `lock` flavors with `perform`?

Reply:
Yes actually -- consider what happens if the lock operation gets cancelled. In that case the perform action will be consumed without being run, which can result in cancel safety issues of its own.

Reply:
Added some documentation.

Reply:
I read the comment, but it just says that the situation would be cancel unsafe. I'm not sure how the scenario is different than if I did
```rust
mutex.lock().await.unwrap().perform(|data| ..);
```